src/geom/SkyWcs.cc

```diff
@@ -0,0 +1,249 @@
+                                    square(skyDx[0] * skyDy[1] - skyDx[1] * skyDy[0]));
+    return std::sqrt(pixAreaRadSq) * radians;
+}
+
+Point2D SkyWcs::getPixelOrigin() const { return this->tranInverse(getSkyOrigin()); }
+
+SpherePoint SkyWcs::getSkyOrigin() const {
+    // CRVAL is stored as the SkyRef property of the sky frame (the current frame of the SkyWcs)
+    auto const skyFrame = detail::getSkyFrame(*_frameSet, ast::FrameSet::CURRENT, false);
+    auto const crvalRad = skyFrame->getSkyRef();
+    return SpherePoint(crvalRad.data());
+}
+
+Eigen::Matrix2d SkyWcs::getCdMatrix() const {
+    // Use a copy because looking up the IWC frame alters the FrameSet
+    auto frameSet = getFrameSet()->copy();
+    auto prototype = ast::Frame(2, "Domain=IWC");
+    auto iwcFrameSet = frameSet->findFrame(prototype);
+    if (!iwcFrameSet) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "IWC frame not found");
+    }
+    auto iwcTransform = Transform<Point2Endpoint, Point2Endpoint>(*iwcFrameSet);
+    return iwcTransform.getJacobian(getPixelOrigin());
+}
+
+SkyWcs SkyWcs::getTanWcs(Point2D const& pixel) const {
+    // I assume the CD matrix is independent of pixel position;
+    // if we find a case where this is not so, add a version of getCdMatrix that accepts pixel position
+    auto const crval = tranForward(pixel);
+    auto const cdMatrix = getCdMatrix();
+    return SkyWcs(pixel, crval, cdMatrix);
+}
+
+SkyWcs SkyWcs::shiftedPixelOrigin(double dx, double dy) const {
+    auto frameSet = getFrameSet()->copy();
+
+    // Save the SkyFrame so we can search for it at the end, to make it current again
+    auto skyFrameIndex = frameSet->getCurrent();
+
+    // Find the GRID frame
+    if (!frameSet->findFrame(ast::Frame(2, "Domain=GRID"))) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "GRID frame not found");
+    }
+    auto gridFrameIndex = frameSet->getCurrent();
+
+    // Find the PIXEL0 frame and the old mapping from GRID to PIXEL0
+    frameSet->setBase(gridFrameIndex);
+    auto oldGridToPixel0Map = frameSet->findFrame(ast::Frame(2, "Domain=PIXEL0"));
+    if (!oldGridToPixel0Map) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "PIXEL0 frame not found");
+    }
+    auto oldPixel0Index = frameSet->getCurrent();
+    auto pixel0Frame = frameSet->getFrame(oldPixel0Index);
+
+    // Look for the optional ACTUAL_PIXEL0 frame, and the mapping from PIXEL0 to ACTUAL_PIXEL0
+    frameSet->setBase(oldPixel0Index);
+    auto oldActualPixel0Index = ast::FrameSet::NOFRAME;
+    std::shared_ptr<ast::Frame> actualPixel0Frame;
+    auto pixelToActualPixelMap = frameSet->findFrame(ast::Frame(2, "Domain=ACTUAL_PIXEL0"));
+    if (pixelToActualPixelMap) {
+        actualPixel0Frame = frameSet->getFrame(ast::FrameSet::CURRENT);
+        oldActualPixel0Index = frameSet->getCurrent();
+    }
+
+    // Compute new mapping from GRID to PIXEL0
+    std::vector<double> shiftVec = {dx, dy};
```

Comment:
But `shift` is now the name of the `Extent2D` argument, so I'm retaining `shiftVec`. I know you hate Hungarian notation but I find that in very small doses it can sometimes be helpful for disambiguating variable names, and I think this is precisely that sort of situation.